components/stratum/stratum_api.c

```diff
@@ -10,15 +10,80 @@
 #include "esp_ota_ops.h"
 #include "lwip/sockets.h"
 #include "utils.h"
+#include "esp_timer.h"
 #include <stdio.h>
 #include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
 
 #define BUFFER_SIZE 1024
 static const char * TAG = "stratum_api";
 
 static char * json_rpc_buffer = NULL;
 static size_t json_rpc_buffer_size = 0;
 
+static RequestTiming request_timings[MAX_REQUEST_IDS];
+static bool initialized = false;
+
+static void init_request_timings() {
+    if (!initialized) {
+        for (int i = 0; i < MAX_REQUEST_IDS; i++) {
+            request_timings[i].timestamp_us = 0;
+            request_timings[i].tracking = false;
+        }
+        initialized = true;
+    }
+}
+
+static int get_request_id(const char *msg) {
+    cJSON *root = cJSON_Parse(msg);
+    if (!root) return -1;
+    
+    cJSON *id = cJSON_GetObjectItem(root, "id");
+    if (!id || !cJSON_IsNumber(id)) {
+        cJSON_Delete(root);
+        return -1;
+    }
+    
+    int request_id = id->valueint;
+    cJSON_Delete(root);
+    return request_id;
+}
+
+static RequestTiming* get_request_timing(int request_id) {
+    if (request_id < 0 || request_id >= MAX_REQUEST_IDS) return NULL;
+    return &request_timings[request_id];
+}
+
+void STRATUM_V1_stamp_tx(const char *msg)
+{
+    init_request_timings();
+    int request_id = get_request_id(msg);
+    if (request_id >= 0) {
+        RequestTiming *timing = get_request_timing(request_id);
+        if (timing) {
+            timing->timestamp_us = esp_timer_get_time();
+            timing->tracking = true;
+        }
+    }
+}
+
+double STRATUM_V1_get_response_time_ms(const char *msg)
+{
+    init_request_timings();
+    int request_id = get_request_id(msg);
```

Comment:
I think it should be possible to avoid having to parse the message again here as well. Pass in an id and maybe use stratum_api_v1_message from the STRATUM parse method.